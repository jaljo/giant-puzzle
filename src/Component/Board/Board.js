import React from 'react'
import './Board.css'

export default ({
  lines,
  meh,
}) =>
  <main>
    <div data-is="main-board">
      {lines.map((line, idx) =>
        <section data-is="line" key={`line-${idx}`}>
          {line.map((tile, idx) => <Tile {...tile } key={`tile-${idx}`}/>)}
        </section>
      )}
    </div>

    {meh && 'Cant move there, asshat'}
  </main>

const Tile = ({
  x,
  y,
  char,
  locked,
}) =>
  <div data-is="tile" className={`${locked ? 'is-locked' : ''}`}>
    {`(${x}, ${y})`}
    {char && <Character {...char} />}
  </div>

const Character = ({
  image,
}) =>
  <div data-is="character">
    <img src={image} alt="icon" />
  </div>
