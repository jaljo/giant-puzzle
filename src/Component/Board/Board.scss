main {
  div[data-is="main-board"] {
    width: 640px;
    height: 768px;
    display: flex;
    flex-direction: column;
    margin-bottom: 20px;

    section[data-is="line"] {
      flex: 1;
      display: flex;
      flex-direction: row;

      div[data-is="tile"] {
        flex: 1;
        border: 1px black dashed;
        min-height: 128px;
        min-width: 128px;

        &.is-locked {
          background-color: black;
        }

        div[data-is="character"] {
          width: 128px;
          height: 128px;

          &.chick.up {
            background-image: url('/images/Chick_Up.png')
          }

          &.chick.down {
            background-image: url('/images/Chick_Down.png')
          }

          &.chick.left {
            background-image: url('/images/Chick_Left.png')
          }

          &.chick.right {
            background-image: url('/images/Chick_Right.png')
          }

          &.fox.up {
            background-image: url('/images/Fox_Up.png')
          }

          &.fox.down {
            background-image: url('/images/Fox_Down.png')
          }

          &.fox.left {
            background-image: url('/images/Fox_Left.png')
          }

          &.fox.right {
            background-image: url('/images/Fox_Right.png')
          }
        }
      }
    }
  }

  p.game-over {
    color: red;
    font-size: 5rem;
  }

  p.game-won {
    color: green;
    font-size: 5rem;
  }
}
