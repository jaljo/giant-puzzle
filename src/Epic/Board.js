import { fromEvent, zip } from 'rxjs'
import { combineEpics, ofType } from 'redux-observable'
import { map, filter, withLatestFrom, mergeMap } from 'rxjs/operators'
import {
  findTileByCoordinates,
  findTileWithCharacter,
  getWinTileA,
  getWinTileB,
  getOppositeDirection,
} from './../Util'
import {
  __,
  allPass,
  apply,
  complement,
  cond,
  dec,
  filter as rfilter,
  evolve,
  head,
  ifElse,
  inc,
  includes,
  isEmpty,
  isNil,
  map as rmap,
  o,
  omit,
  or,
  path,
  pipe,
  prop,
  replace,
  toLower,
} from 'ramda'
import {
  ARROW_KEY_PRESSED,
  GUARDIAN_REGULAR,
  GUARDIAN_REVERSE,
  MAIN_CHARACTER,
  MOVE_CHARACTER,
  NEXT_COORDINATES_OBTAINED,
  REQUEST_CHARACTER_MOVE,
  arrowKeyPressed,
  gameOver,
  meh,
  moveCharacter,
  nextCoordinatesObtained,
  requestCharacterMove,
  winGame,
} from './../Redux/State/Board'

// isMainCharacterMove :: Action -> Boolean
const isMainCharacterMove = action => action.characterId === MAIN_CHARACTER.id
const isRegularGuardianMove = action => action.characterId === GUARDIAN_REGULAR.id
const isReverseGuardianMove = action => action.characterId === GUARDIAN_REVERSE.id

// isArrowKeyPressed :: [String] -> KeyboardEvent -> Boolean
const isArrowKeyPressed = keyMap => pipe(
  prop('key'),
  includes(__, keyMap),
)

// keyEventToMoveActionEpic :: Epic -> Observable Action ARROW_KEY_PRESSED
const keyEventToMoveActionEpic = (action$, state$, { keyMap }) =>
  fromEvent(window, 'keyup').pipe(
    filter(isArrowKeyPressed(keyMap)),
    withLatestFrom(state$),
    filter(([ _, state ]) => !state.Board.gameOver),
    filter(([ _, state ]) => !state.Board.winGame),
    map(pipe(
      head,
      prop('key'),
      toLower,
      replace('arrow', ''),
      arrowKeyPressed,
    )),
  )

// toLeft :: Tile -> Coordinates
const toLeft = evolve({ x: dec })
const toRight = evolve({ x: inc })
const toUp = evolve({ y: inc })
const toDown = evolve({ y: dec })

// tileToCoordinates :: Maybe Tile -> Maybe Coordinates
const tileToCoordinates = omit(['char', 'locked'])

// directionIs :: String -> [Any, Action] -> Boolean
const directionIs = key => ([ _, action ]) => action.direction === key

// isNotOutOfBounds :: Maybe Tile -> Boolean
const isNotOutOfBounds = tile => tile.x !== null && tile.y !== null

// isNotGuarded :: Tile -> Boolean
const isNotGuarded = pipe(
  path(['char', 'id']),
  complement(includes(__, [GUARDIAN_REGULAR.id, GUARDIAN_REVERSE.id])),
)

// isNotLocked :: Tile -> Boolean
const isNotLocked = complement(prop('locked'))

// tileIsFree :: Tile :: Boolean
const tileIsFree = allPass([isNotOutOfBounds, isNotLocked, isNotGuarded])

// haveNotTheSameDestination :: (Action.NEXT_COORDINATES_OBTAINED, Action.NEXT_COORDINATES_OBTAINED) -> Boolean
const haveNotTheSameDestination = (a, b) => or(
  a.targetTile.x !== b.targetTile.x,
  a.targetTile.y !== b.targetTile.y,
)

// moveCharacterOrMeh :: Action.NEXT_COORDINATES_OBTAINED -> Action.MOVE_CHARACTER Action.MEH
const moveCharacterOrMeh = ifElse(
  o(tileIsFree, prop('targetTile')),
  action => moveCharacter(
    action.characterId,
    action.direction,
    tileToCoordinates(action.targetTile)
  ),
  o(meh, prop('characterId')),
)

// obtainNextCoordinatesEpic :: Epic -> Observable Action NEXT_COORDINATES_OBTAINED
const obtainNextCoordinatesEpic = (action$, state$) =>
  action$.pipe(
    ofType(REQUEST_CHARACTER_MOVE),
    withLatestFrom(state$),
    // find the tile where the main character is
    // Observable [Action, State] -> Observable Tile
    map(([ action, state ]) => findTileWithCharacter(action.characterId)(state.Board.lines)),
    withLatestFrom(action$),
    // compute target coordinates
    // Observable [Tile, Action] -> Observable Coordinates
    map(pipe(
      cond([
        [directionIs('up'), ([ tile ]) => toUp(tile)],
        [directionIs('down'), ([ tile ]) => toDown(tile)],
        [directionIs('left'), ([ tile ]) => toLeft(tile)],
        [directionIs('right'), ([ tile ]) => toRight(tile)],
      ]),
      tileToCoordinates,
    )),
    withLatestFrom(state$),
    // find the tile matching the target coordinates (if any)
    // Observable [Coordinates, State] -> Observable Tile
    map(([ coordinates, state ]) => findTileByCoordinates(coordinates)(state.Board.lines)),
    withLatestFrom(action$),
    map(([ tile, action ]) => nextCoordinatesObtained(
      action.characterId,
      action.direction,
      tile,
    )),
  )

// requestMainCharacterMoveEpic :: Epic -> Observable Action REQUEST_CHARACTER_MOVE
const requestMainCharacterMoveEpic = action$ =>
  action$.pipe(
    ofType(ARROW_KEY_PRESSED),
    // @TODO inject characters id rather than using them directly
    map(action => requestCharacterMove(MAIN_CHARACTER.id, action.direction)),
  )

// moveMainCharacterEpic :: Epic -> Observable Action MOVE_CHARACTER MEH
const moveMainCharacterEpic = action$ =>
  action$.pipe(
    ofType(NEXT_COORDINATES_OBTAINED),
    filter(isMainCharacterMove),
    map(moveCharacterOrMeh),
  )

// requestRegularGuardianMoveEpic :: Epic -> REQUEST_CHARACTER_MOVE
const requestRegularGuardianMoveEpic = action$ =>
  action$.pipe(
    ofType(MOVE_CHARACTER),
    filter(isMainCharacterMove),
    // @TODO inject characters id rather than using them directly
    map(action => requestCharacterMove(GUARDIAN_REGULAR.id, action.direction)),
  )

// requestReverseGuardianMoveEpic :: Epic -> Observable Action REQUEST_CHARACTER_MOVE
const requestReverseGuardianMoveEpic = action$ =>
  action$.pipe(
    ofType(MOVE_CHARACTER),
    filter(isMainCharacterMove),
    map(pipe(
      prop('direction'),
      getOppositeDirection,
      // @TODO inject characters id rather than using them directly
      direction => requestCharacterMove(GUARDIAN_REVERSE.id, direction),
    )),
  )

// moveGuardianEpic :: Epic -> [Observable Action *]
const moveGuardianEpic = action$ =>
  zip(
    action$.pipe(
      ofType(NEXT_COORDINATES_OBTAINED),
      filter(isRegularGuardianMove),
    ),
    action$.pipe(
      ofType(NEXT_COORDINATES_OBTAINED),
      filter(isReverseGuardianMove),
    ),
  ).pipe(
    filter(apply(haveNotTheSameDestination)),
    mergeMap(rmap(moveCharacterOrMeh)),
  )

// gameOverEpic :: Epic -> Observable Action GAME_OVER
const gameOverEpic = (action$, state$) =>
  action$.pipe(
    ofType(MOVE_CHARACTER),
    withLatestFrom(state$),
    map(([ _, state ]) => findTileWithCharacter(MAIN_CHARACTER.id)(state.Board.lines)),
    filter(isNil),
    map(gameOver),
  )

// charIsGuardian :: Tile -> Boolean
const charIsGuardian = pipe(
  path(['char', 'id']),
  includes(__, [GUARDIAN_REGULAR.id, GUARDIAN_REVERSE.id]),
)

// everyTileIsGuarded :: [Tile] -> Boolean
export const everyTileIsGuarded = pipe(
  rfilter(complement(charIsGuardian)),
  isEmpty,
)

// winGameEpic :: Epic -> Observable Action WIN_GAME
const winGameEpic = (action$, state$) =>
    action$.pipe(
      ofType(MOVE_CHARACTER),
      withLatestFrom(state$),
      map(([ _, state ]) => [
        getWinTileA(state.Board.lines),
        getWinTileB(state.Board.lines),
      ]),
      filter(everyTileIsGuarded),
      map(winGame),
    )

export default combineEpics(
  gameOverEpic,
  keyEventToMoveActionEpic,
  moveGuardianEpic,
  moveMainCharacterEpic,
  obtainNextCoordinatesEpic,
  requestMainCharacterMoveEpic,
  requestRegularGuardianMoveEpic,
  requestReverseGuardianMoveEpic,
  winGameEpic,
)
