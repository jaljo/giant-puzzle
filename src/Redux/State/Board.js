import { createReducer } from '../../Util'
import { find, propEq } from 'ramda'

export const MAIN_CHARACTER = {
  id: 'main-character',
  image: 'https://image.flaticon.com/icons/svg/2754/2754522.svg',
}

export const GUARDIAN_REGULAR = {
  id: 'guardian-regular',
  image: 'https://image.flaticon.com/icons/svg/562/562802.svg',
}

export const GUARDIAN_REVERSE = {
  id: 'guardian-reverse',
  image: 'https://image.flaticon.com/icons/svg/2699/2699064.svg',
}

const CHARACTER_MAP = [
  MAIN_CHARACTER,
  GUARDIAN_REGULAR,
  GUARDIAN_REVERSE,
]

const findCharacter = id => find(propEq('id', id), CHARACTER_MAP)

export const INITIAL_STATE = {
  meh: false,
  gameOver: false,
  winGame: false,
  lines: [
    // l6
    [
      {
        x: 0,
        y: 5,
        char: null,
        locked: false,
      }, {
        x: 1,
        y: 5 ,
        char: null,
        locked: false,
      }, {
        x: 2,
        y: 5,
        char: null,
        locked: true,
      }, {
        x: 3,
        y: 5 ,
        char: null,
        locked: false,
      }, {
        x: 4,
        y: 5,
        char: null,
        locked: false,
      }
    ],
    // l5
    [
      {
        x: 0,
        y: 4,
        char: null,
        locked: false,
      }, {
        x: 1,
        y: 4 ,
        char: null,
        locked: false,
      }, {
        x: 2,
        y: 4,
        char: GUARDIAN_REVERSE,
        locked: false,
      }, {
        x: 3,
        y: 4 ,
        char: null,
        locked: false,
      }, {
        x: 4,
        y: 4,
        char: null,
        locked: false,
      }
    ],
    // l4
    [
      {
        x: 0,
        y: 3,
        char: null,
        locked: false,
      }, {
        x: 1,
        y: 3 ,
        char: null,
        locked: false,
      }, {
        x: 2,
        y: 3,
        char: null,
        locked: false,
      }, {
        x: 3,
        y: 3 ,
        char: null,
        locked: false,
      }, {
        x: 4,
        y: 3,
        char: null,
        locked: false,
      }
    ],
    // l3
    [
        {
          x: 0,
          y: 2,
          char: null,
          locked: true,
        }, {
          x: 1,
          y: 2 ,
          char: null,
          locked: false,
        }, {
          x: 2,
          y: 2,
          char: MAIN_CHARACTER,
          locked: false,
        }, {
          x: 3,
          y: 2 ,
          char: null,
          locked: false,
        }, {
          x: 4,
          y: 2,
          char: null,
          locked: true,
        }
    ],
    // l2
    [
        {
          x: 0,
          y: 1,
          char: null,
          locked: true,
        }, {
          x: 1,
          y: 1,
          char: null,
          locked: false,
        }, {
          x: 2,
          y: 1,
          char: null,
          locked: false,
        }, {
          x: 3,
          y: 1 ,
          char: null,
          locked: false,
        }, {
          x: 4,
          y: 1,
          char: null,
          locked: true,
        }
    ],
    // l1
    [
        {
          x: 0,
          y: 0,
          char: null,
          locked: true,
        }, {
          x: 1,
          y: 0,
          char: null,
          locked: true,
        }, {
          x: 2,
          y: 0,
          char: GUARDIAN_REGULAR,
          locked: false,
        }, {
          x: 3,
          y: 0,
          char: null,
          locked: true,
        }, {
          x: 4,
          y: 0,
          char: null,
          locked: true,
        }
    ],
  ]
}

export const ARROW_KEY_PRESSED = '@giant-puzzle/Board/ARROW_KEY_PRESSED'
export const REQUEST_CHARACTER_MOVE = '@giant-puzzle/Board/REQUEST_CHARACTER_MOVE'
export const NEXT_COORDINATES_OBTAINED = '@giant-puzzle/Board/NEXT_COORDINATES_OBTAINED'
export const MOVE_CHARACTER = '@giant-puzzle/Board/MOVE_CHARACTER'
export const MEH = '@giant-puzzle/Board/MEH'
export const GAME_OVER = '@giant-puzzle/Board/GAME_OVER'
export const WIN_GAME = '@giant-puzzle/Board/WIN_GAME'
export const RETRY = '@giant-puzzle/Board/RETRY'

// arrowKeyPressed :: direction
export const arrowKeyPressed = direction => ({
  type: ARROW_KEY_PRESSED,
  direction,
})

// requestCharacterMove :: (String, String) -> Action
export const requestCharacterMove = (characterId, direction) => ({
  type: REQUEST_CHARACTER_MOVE,
  characterId,
  direction,
})

// nextCoordinatesObtained :: (String, String, Tile) -> Action
export const nextCoordinatesObtained = (characterId, direction, targetTile) => ({
  type: NEXT_COORDINATES_OBTAINED,
  characterId,
  direction,
  targetTile,
})

// moveCharacter :: (String, String -> Coordinates) -> Action
export const moveCharacter = (characterId, direction, coordinates) => ({
  type: MOVE_CHARACTER,
  characterId,
  direction,
  coordinates,
})

// meh :: String -> Action
export const meh = characterId => ({
  type: MEH,
  characterId,
})

// gameOver :: () -> Action
export const gameOver = () => ({ type: GAME_OVER })

// winGame :: () -> Action
export const winGame = () => ({ type: WIN_GAME })

// retry :: () -> Action
export const retry = () => ({ type: RETRY })

export default createReducer(INITIAL_STATE, {
  [ARROW_KEY_PRESSED]: state => ({
    ...state,
    meh: false,
  }),

  [MOVE_CHARACTER]: (state, { characterId, coordinates }) => ({
    ...state,
    lines: state.lines.map(
      line => line.map(tile => ({
        ...tile,
        char: resolveCharacter(tile, coordinates, characterId),
      }))
    ),
  }),

  [MEH]: state => ({
    ...state,
    meh: true,
  }),

  [GAME_OVER]: state => ({
    ...state,
    gameOver: true,
  }),

  [WIN_GAME]: state => ({
    ...state,
    winGame: true,
  }),

  [RETRY]: () => INITIAL_STATE,
})

// move character on the target slide, remove it from the initial slide
// dont do anything for tiles with other characters on
//
// resolveCharacter :: (Tile, Coordinates, String) -> Maybe Character
export const resolveCharacter = (tile, coordinates, characterId) =>
  (tile.x === coordinates.x && tile.y === coordinates.y)
    ? findCharacter(characterId)
    : (tile.char && tile.char.id === characterId)
      ? null
      : tile.char
