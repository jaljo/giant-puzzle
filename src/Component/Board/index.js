import Board from './Board'
import { connect } from 'react-redux'

// mapStateToProps :: State -> Props
const mapStateToProps = state => ({
  lines: state.Board,
})

// Board :: Props -> React.Component
export default connect(
  mapStateToProps,
)(Board)
