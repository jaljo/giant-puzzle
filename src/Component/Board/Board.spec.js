import React from 'react'
import renderer from 'react-test-renderer'
import Board from './Board'

describe('Component/Board', () => {
  it('renders a board with an empty tile and a tile with a character', () => {
    const mockChar = {
      asset: 'fox',
      direction: 'down',
    }

    const mockLines = [
      [
        { x: 0, y: 0, char: null },
        { x: 0, y: 0, char: mockChar },
      ]
    ]

    expect(
      renderer.create(<Board lines={mockLines} /> )
    ).toMatchSnapshot()
  })
})
