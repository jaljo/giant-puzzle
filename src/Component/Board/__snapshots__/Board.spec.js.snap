// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component :: Board renders a board with an empty tile and a tile with a character 1`] = `
<div
  data-is="main-board"
>
  <section
    data-is="line"
  >
    <div
      data-is="tile"
    />
    <div
      data-is="tile"
    >
      <div
        className="fox down "
        data-is="character"
      />
    </div>
  </section>
</div>
`;
